### PR TITLE
AS-1920: fix IndexError due to inconsistent returned value format

### PR DESCRIPTION
**Pull request by mcamarad**
## Why?

A bug caused by inconsistencies between the formats returned from the different functions in the `get_location_info` function caused the `double_check_countries` and `update_mapping_countries` to fail with IndexErrors. This PR addresses that by making the format consistent and changing all returning values from `[{}]` to `[]`. 

Additionally, it addresses a bug in the blacklist loading that caused the blacklist to not work as expected. The error was due to the keywords not being split correctly.

## What?

- Changes all returning values from `[{}]` to `[]` in `geocoder.py`
- Updates tests to support the change
- Fix blacklist issue

### CODE REVIEW

File: geocoder_module/blacklist.txt [hunk 1]
```diff
@@ -29,14 +29,6 @@ mom
 dad
```
Comment (s0rthak) [inline geocoder_module/blacklist.txt:29]: Any reason we're using a txt file for blacklist?
Can we make it a json file instead so reading and using it is easier?

File: geocoder_module/geocoder.py [hunk 2]
```diff
@@ -47,7 +47,9 @@ def __init__(self) -> None:
                     "blacklist.txt",
                 ),
                 "r",
-            ).readlines(),
+            )
+            .read()
+            .splitlines(),
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:52]: My concern with txt files is that it is possible a single line can have extra whitespace which may cause the blacklist to not work currently for each keyword

File: geocoder_module/geocoder.py [hunk 3]
```diff
@@ -326,6 +327,11 @@ def _validate_locations(
 
         """
         validated_results = []
+        if initial_results == []:
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:330]: You can use `if not initial_results` to simplify this. I personally use `if len(initial_results) == 0` to highlight that it's a list I am comparing. Either is fine.

File: geocoder_module/geocoder.py [hunk 4]
```diff
@@ -326,6 +327,11 @@ def _validate_locations(
 
         """
         validated_results = []
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:329]: Can we put `validated_results = []` after the check for empty initial_results?

File: geocoder_module/geocoder.py [hunk 5]
```diff
@@ -343,7 +349,7 @@ def _validate_locations(
                 else:
                     continue
             if not validated_results:
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:351]: So we return empty result if the validation fails for any location in the `initial_results` list?

File: geocoder_module/geocoder.py [hunk 6]
```diff
@@ -343,7 +349,7 @@ def _validate_locations(
...
-                validated_results = [{}]
+                validated_results = []
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:352]: Do you need to do add this line?

File: geocoder_module/geocoder.py [hunk 7]
```diff
@@ -531,7 +537,7 @@ def update_mapping_countries(
                 name, country=new_country, best_matching=True
             )
             # check if the reference country can be used for this location
-            if new_location != [{}]:
+            if new_location:
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:540]: This is clean!

**Comment by ericanand** [comment]
I see when I search for `[{}]` in the code, that we get mentions of this being used in the test cases for the functions `get_location_from_coordinates` and `update_country_for_locations`. Do we need to update these functions too to not return `[{}]`?

File: geocoder_module/geocoder.py [hunk 8]
```diff
@@ -326,6 +327,11 @@ def _validate_locations(
 
         """
         validated_results = []
+        if initial_results == []:
+            logging.warn(
+                f"Can't validate location {location}. Empty Geocoder hits: {initial_results}"
+            )
+            return []
```
Comment (ericanand) [inline geocoder_module/geocoder.py:334]: Returning early! Nice.

File: geocoder_module/geocoder.py [hunk 9]
```diff
@@ -598,11 +604,11 @@ def double_check_countries(
             logging.info(
                 "Location Edge Case 0 detected: Only one country detected in event locations"
             )
-            if not locations[0] or not "name" in locations[0]:
+            if not locations or not "name" in locations[0]:
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:607]: When there are two nots inside an if, it's better to group them in a single not like this so it's more readable:
`if not (len(locations) > 0 and "name" in locations[0])`

File: tests/test_get_location.py [hunk 10]
```diff
@@ -338,17 +338,50 @@ def test_get_location_info_returns_empty_list_when_location_found_by_geocoder_ca
...
+            "latitude": "-33.86778",
+            "longitude": "151.20844",
+            "country": "Australia",
+            "continent": "Oceania",
+        }
+        mock_get.return_value.json = expected_get_geonames_api_output
+        response = geocoder._validate_locations(
+            [{"name": "Sydney", "country": "Australia"}], "Sydney"
+        )
+        assert response == expected_output
```
Comment (ericanand) [inline tests/test_get_location.py:362]: Is this expected output just checking that `[] == []`?

File: tests/test_get_location.py [hunk 11]
```diff
@@ -338,17 +338,50 @@ def test_get_location_info_returns_empty_list_when_location_found_by_geocoder_ca
...
             expected_get_geonames_api_output_2,
         ]
-        expected_output = [{}]
+        expected_output = []
         response = geocoder.get_location_info("asia petroleum hub")
         assert mock_get.called
         assert response == expected_output
 
 
+class TestValidateLocations:
```
Comment (ericanand) [inline tests/test_get_location.py:347]: Love me some tests 👍

Comment (s0rthak) [inline tests/test_get_location.py:347]: Good job on using class! They make our tests much cleaner

**Review by s0rthak — COMMENTED** [review]
This is the first time I'm reading the code so I'm being very nit-picky. Apologies in advance

**Review by ericanand — COMMENTED** [review]
I'd like to test this out if possible - maybe we could have this run on the sample creation service locally (without updating elasticsearch), and see if it works as expected?

**Comment by s0rthak** [comment]
image
We'll also need a fix in sample-creation service for this I think

**Comment by ericanand** [comment]
> image We'll also need a fix in sample-creation service for this I think

Great spot! Yes we will need to raise a parallel PR for that 🙏

**Comment by mcamarad** [comment]
Thanks both for the great comments! You are correct with the `samples-creation` PRs. I was already working on that PR. I forgot to mention it in the Additional Info

File: geocoder_module/geocoder.py (hunk 5 above)
Comment (mcamarad) [reply geocoder_module/geocoder.py:351]: Yes, the current config only returns one result, but if we changed that to multiple results for a given location, we would then return empty results for any individual `initial_results` that are not in the Geonames.

File: geocoder_module/geocoder.py (hunk 6 above)
Comment (mcamarad) [reply geocoder_module/geocoder.py:352]: Well spotted, actually we don't need it.

File: geocoder_module/geocoder.py (hunk 2 above)
Comment (mcamarad) [reply geocoder_module/geocoder.py:52]: I will change it to a JSON

File: geocoder_module/geocoder.py (hunk 9 above)
Comment (mcamarad) [reply geocoder_module/geocoder.py:607]: If I group them then it doesn't work as expected. Each clause needs to have its own not
image

File: tests/test_get_location.py (hunk 10 above)
Comment (mcamarad) [reply tests/test_get_location.py:362]: Apologies, that was a mistake

**Comment by mcamarad** [comment]
@s0rthak @ericanand Ready for re-review! 🚀

File: geocoder_module/blacklist.txt (hunk 1 above)
Comment (mcamarad) [reply geocoder_module/blacklist.txt]: Changed to JSON

File: geocoder_module/blacklist.json [hunk 12]
```diff
@@ -0,0 +1,50 @@
+[{"data": ["america",
```
Comment (s0rthak) [inline geocoder_module/blacklist.json:1]: We can put a simple array as well. Don't need to nest it under data

File: geocoder_module/geocoder.py [hunk 13]
```diff
@@ -40,16 +40,26 @@ def __init__(self) -> None:
...
-                    "blacklist.txt",
-                ),
-                "r",
-            ).readlines(),
+            "blacklist_path": "blacklist.json",
         }
 
+        try:
+            self.blacklist = json.load(
+                open(
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:48]: This is v complex 1 liner. Can we make it more readable?
Also with i/o in python, we should use with... to make sure the file is closed automatically

Example:
with open("test.txt", encoding = 'utf-8') as f:
   // perform file operations

**Review by s0rthak — COMMENTED** [review]
Some more comments on loading the blacklist file

File: geocoder_module/blacklist.json (hunk 12 above)
Comment (mcamarad) [reply geocoder_module/blacklist.json:1]: Got it! Thanks!

File: geocoder_module/geocoder.py (hunk 13 above)
Comment (mcamarad) [reply geocoder_module/geocoder.py]: Still looks like a bit of a monstrous piece of code, but I tried to make it more readable.

**Comment by mcamarad** [comment]
@s0rthak @ericanand Ready for Round 2.5! 🚀

File: geocoder_module/geocoder.py [hunk 14]
```diff
@@ -40,82 +40,82 @@ def __init__(self) -> None:
...
-            self.map_country_neighbors = json.load(
-                open(
-                    os.path.join(
-                        os.path.dirname(os.path.dirname(geocoder_module.__file__)),
-                        "geocoder_module",
-                        self.config["country_neighbors_path"],
-                    ),
-                    "r",
-                )
+            blacklist_path = os.path.join(
```
Comment (s0rthak) [inline geocoder_module/geocoder.py:47]: You can generalize loading of json files in a function like this:

```
def load_json_file(file_name: str):
    try:
        file_path = os.path.join(
            os.path.dirname(os.path.dirname(geocoder_module.__file__)),
            "geocoder_module",
            file_name,
        )
        with open(file_path, "r", encoding="utf-8") as f:
            data = json.load(f)
            return data
    except Exception as error:
        logging.error(f"{error} The json file {file_name} specified in the configuration can't be read.")
```

**Review by s0rthak — COMMENTED** [review]
Just a comment about being DRY with loading jsons. Rest looks good to me code-wise.

File: geocoder_module/geocoder.py (hunk 14 above)
Comment (mcamarad) [reply geocoder_module/geocoder.py]: Thanks a lot!

**Review by ericanand — APPROVED** [review]
Great stuff 👍

**Pull request merged**